File: core/src/test/scala/unit/kafka/log/remote/RemoteLogReaderTest.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.log.remote

import java.nio.file.Files
import java.util.{Optional, Properties}
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.ReentrantReadWriteLock
import java.util.concurrent.{CompletableFuture, RejectedExecutionException}

import kafka.server.{BrokerTopicStats, FetchDataInfo, FetchTxnCommitted, LogOffsetMetadata, RemoteStorageFetchInfo}
import org.apache.kafka.common.TopicPartition
import org.apache.kafka.common.config.AbstractConfig
import org.apache.kafka.common.errors.OffsetOutOfRangeException
import org.apache.kafka.common.record.{CompressionType, MemoryRecords, SimpleRecord}
import org.apache.kafka.common.requests.FetchRequest.PartitionData
import org.apache.kafka.common.utils.SystemTime
import org.apache.kafka.server.log.remote.storage.RemoteLogManagerConfig
import org.junit.jupiter.api.Assertions.{assertEquals, assertThrows, assertTrue}
import org.junit.jupiter.api.Test

import scala.jdk.CollectionConverters._

class RemoteLogReaderTest {

  @Test
  def testReadRemoteLog(): Unit = {
    val rlm = new MockRemoteLogManager(2, 10)
    val tp = new TopicPartition("test", 0)
    val fetchInfo = new PartitionData(200, 200, 1000, Optional.of(1))
    val resultFuture = new CompletableFuture[RemoteLogReadResult]
    def callback(result: RemoteLogReadResult): Unit = {
      resultFuture.complete(result)
    }
    rlm.asyncRead(RemoteStorageFetchInfo(1000, minOneMessage = true, tp, fetchInfo, FetchTxnCommitted), callback)
    Thread.sleep(100)

    val dataInfo = resultFuture.get.info.get
    assertEquals(None, resultFuture.get.error)
    assertEquals(200, dataInfo.fetchOffsetMetadata.messageOffset)
    assertEquals(2, dataInfo.records.records.asScala.size)
    assertEquals(202, dataInfo.records.records.iterator.next.offset)
  }

  @Test
  def testTaskQueueFullAndCancelTask(): Unit = {
    val rlm = new MockRemoteLogManager(5, 20)
    rlm.pause()

    val tp = new TopicPartition("test", 0)
    val tasks = new Array[RemoteLogManager#AsyncReadTask](26)
    val finishedTasks = Array.fill[Boolean](26)(false)
    val finishCount = new AtomicInteger(0)
    def callback(result: RemoteLogReadResult): Unit = {
      assertEquals(None, result.error)
      finishCount.addAndGet(1)
      finishedTasks(result.info.get.fetchOffsetMetadata.messageOffset.asInstanceOf[Int]) = true
    }

    for (i <- 0 to 24) {
      val fetchInfo = new PartitionData(i, 0, 1000, Optional.of(1))
      val future = rlm.asyncRead(RemoteStorageFetchInfo(1000, minOneMessage = true, tp, fetchInfo, FetchTxnCommitted), callback)
      tasks(i) = future
    }
    assertEquals(0, finishCount.get)

    assertThrows(classOf[RejectedExecutionException], () => {
      val fetchInfo = new PartitionData(25, 0, 1000, Optional.of(1))
      rlm.asyncRead(RemoteStorageFetchInfo(1000, minOneMessage = true, tp, fetchInfo, FetchTxnCommitted), callback)
    })

    tasks(7).cancel(false)
    tasks(10).cancel(true)

    val fetchInfo = new PartitionData(25, 0, 1000, Optional.of(1))
    rlm.asyncRead(RemoteStorageFetchInfo(1000, minOneMessage = true, tp, fetchInfo, FetchTxnCommitted), callback)
    rlm.resume()

    Thread.sleep(200)
    assertEquals(24, finishCount.get)
    assertEquals(24, finishedTasks.count(a => a))
    assertEquals(false, finishedTasks(7))
    assertEquals(false, finishedTasks(10))
  }

  @Test
  def testErr(): Unit = {
    val rlm = new MockRemoteLogManager(2, 10) {
      override def read(remoteStorageFetchInfo: RemoteStorageFetchInfo): FetchDataInfo = {
        throw new OffsetOutOfRangeException("Offset: %d is out of range"
          .format(remoteStorageFetchInfo.fetchInfo.fetchOffset))
      }
    }
    val tp = new TopicPartition("test", 1)
    val fetchInfo = new PartitionData(10000, 0, 1000, Optional.of(1))
    val resultFuture = new CompletableFuture[RemoteLogReadResult]
    def callback(result: RemoteLogReadResult): Unit = {
      resultFuture.complete(result)
    }

    val task = rlm.asyncRead(RemoteStorageFetchInfo(1000, minOneMessage = true, tp, fetchInfo, FetchTxnCommitted), callback)
    Thread.sleep(100)
    assertTrue(task.isDone)
    assertEquals(None, resultFuture.get.info)
    assertEquals(classOf[OffsetOutOfRangeException], resultFuture.get.error.get.getClass)
  }
}

class MockRemoteLogManager(threads: Int, taskQueueSize: Int)
  extends RemoteLogManager(
    _ => None,
    (_, _) => {},
    MockRemoteLogManager.rlmConfig(threads, taskQueueSize),
    new SystemTime,
    1,
    "mock-cluster-id",
    Files.createTempDirectory("kafka-test-").toString,
    new BrokerTopicStats) {

  private val lock = new ReentrantReadWriteLock

  override def read(remoteStorageFetchInfo: RemoteStorageFetchInfo): FetchDataInfo = {
    lock.readLock.lock()
    try {
      val fetchInfo = remoteStorageFetchInfo.fetchInfo
      val recordsArray = Array(
        new SimpleRecord("k1".getBytes, "v1".getBytes),
        new SimpleRecord("k2".getBytes, "v2".getBytes)
      )
      val records = MemoryRecords.withRecords(fetchInfo.fetchOffset + 2, CompressionType.NONE, 1, recordsArray: _*)
      FetchDataInfo(new LogOffsetMetadata(fetchInfo.fetchOffset), records)
    } finally {
      lock.readLock.unlock()
    }
  }

  def pause(): Unit = {
    lock.writeLock.lock()
  }

  def resume(): Unit = {
    lock.writeLock.unlock()
  }
}

object MockRemoteLogManager {

  def rlmConfig(threads: Int, taskQueueSize: Int): RemoteLogManagerConfig = {
    val props = new Properties
    props.put(RemoteLogManagerConfig.REMOTE_LOG_STORAGE_SYSTEM_ENABLE_PROP, true.toString)
    props.put(RemoteLogManagerConfig.REMOTE_LOG_METADATA_MANAGER_CONFIG_PREFIX_PROP, "rlmm.config.")
    props.put(RemoteLogManagerConfig.REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_PROP, "rsm.config.")
    props.put(RemoteLogManagerConfig.REMOTE_STORAGE_MANAGER_CLASS_NAME_PROP, "kafka.log.remote.MockRemoteStorageManager")
    props.put(RemoteLogManagerConfig.REMOTE_LOG_METADATA_MANAGER_CLASS_NAME_PROP, "kafka.log.remote.MockRemoteLogMetadataManager")
    props.put(RemoteLogManagerConfig.REMOTE_LOG_READER_THREADS_PROP, threads.toString)
    props.put(RemoteLogManagerConfig.REMOTE_LOG_READER_MAX_PENDING_TASKS_PROP, taskQueueSize.toString)
    val config = new AbstractConfig(RemoteLogManagerConfig.CONFIG_DEF, props, false)
    new RemoteLogManagerConfig(config)
  }
}

File: core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package kafka.log.remote;

import kafka.utils.TestUtils;

import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.record.Records;
import org.apache.kafka.server.log.remote.quota.RLMQuotaManager;
import org.apache.kafka.server.log.remote.storage.RemoteStorageException;
import org.apache.kafka.storage.internals.log.FetchDataInfo;
import org.apache.kafka.storage.internals.log.LogOffsetMetadata;
import org.apache.kafka.storage.internals.log.RemoteLogReadResult;
import org.apache.kafka.storage.internals.log.RemoteStorageFetchInfo;
import org.apache.kafka.storage.log.metrics.BrokerTopicStats;

import com.yammer.metrics.core.Timer;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;

import java.io.IOException;
import java.util.concurrent.Callable;
import java.util.function.Consumer;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class RemoteLogReaderTest {
    public static final String TOPIC = "test";
    RemoteLogManager mockRLM = mock(RemoteLogManager.class);
    BrokerTopicStats brokerTopicStats = null;
    RLMQuotaManager mockQuotaManager = mock(RLMQuotaManager.class);
    LogOffsetMetadata logOffsetMetadata = new LogOffsetMetadata(100);
    Records records = mock(Records.class);
    Timer timer = mock(Timer.class);

    @BeforeEach
    public void setUp() throws Exception {
        TestUtils.clearYammerMetrics();
        brokerTopicStats = new BrokerTopicStats(true);
        when(timer.time(any(Callable.class))).thenAnswer(ans -> ans.getArgument(0, Callable.class).call());
    }

    @Test
    public void testRemoteLogReaderWithoutError() throws RemoteStorageException, IOException {
        FetchDataInfo fetchDataInfo = new FetchDataInfo(logOffsetMetadata, records);
        when(records.sizeInBytes()).thenReturn(100);
        when(mockRLM.read(any(RemoteStorageFetchInfo.class))).thenReturn(fetchDataInfo);

        Consumer<RemoteLogReadResult> callback = mock(Consumer.class);
        RemoteStorageFetchInfo remoteStorageFetchInfo = new RemoteStorageFetchInfo(0, false, new TopicPartition(TOPIC, 0), null, null, false);
        RemoteLogReader remoteLogReader =
                new RemoteLogReader(remoteStorageFetchInfo, mockRLM, callback, brokerTopicStats, mockQuotaManager, timer);
        remoteLogReader.call();

        // verify the callback did get invoked with the expected remoteLogReadResult
        ArgumentCaptor<RemoteLogReadResult> remoteLogReadResultArg = ArgumentCaptor.forClass(RemoteLogReadResult.class);
        verify(callback, times(1)).accept(remoteLogReadResultArg.capture());
        RemoteLogReadResult actualRemoteLogReadResult = remoteLogReadResultArg.getValue();
        assertFalse(actualRemoteLogReadResult.error.isPresent());
        assertTrue(actualRemoteLogReadResult.fetchDataInfo.isPresent());
        assertEquals(fetchDataInfo, actualRemoteLogReadResult.fetchDataInfo.get());

        // verify the record method on quota manager was called with the expected value
        ArgumentCaptor<Double> recordedArg = ArgumentCaptor.forClass(Double.class);
        verify(mockQuotaManager, times(1)).record(recordedArg.capture());
        assertEquals(100, recordedArg.getValue());

        // Verify metrics for remote reads are updated correctly
        assertEquals(1, brokerTopicStats.topicStats(TOPIC).remoteFetchRequestRate().count());
        assertEquals(100, brokerTopicStats.topicStats(TOPIC).remoteFetchBytesRate().count());
        assertEquals(0, brokerTopicStats.topicStats(TOPIC).failedRemoteFetchRequestRate().count());
        // Verify aggregate metrics
        assertEquals(1, brokerTopicStats.allTopicsStats().remoteFetchRequestRate().count());
        assertEquals(100, brokerTopicStats.allTopicsStats().remoteFetchBytesRate().count());
        assertEquals(0, brokerTopicStats.allTopicsStats().failedRemoteFetchRequestRate().count());
    }

    @Test
    public void testRemoteLogReaderWithError() throws RemoteStorageException, IOException {
        when(mockRLM.read(any(RemoteStorageFetchInfo.class))).thenThrow(new RuntimeException("error"));

        Consumer<RemoteLogReadResult> callback = mock(Consumer.class);
        RemoteStorageFetchInfo remoteStorageFetchInfo = new RemoteStorageFetchInfo(0, false, new TopicPartition(TOPIC, 0), null, null, false);
        RemoteLogReader remoteLogReader =
                new RemoteLogReader(remoteStorageFetchInfo, mockRLM, callback, brokerTopicStats, mockQuotaManager, timer);
        remoteLogReader.call();

        // verify the callback did get invoked with the expected remoteLogReadResult
        ArgumentCaptor<RemoteLogReadResult> remoteLogReadResultArg = ArgumentCaptor.forClass(RemoteLogReadResult.class);
        verify(callback, times(1)).accept(remoteLogReadResultArg.capture());
        RemoteLogReadResult actualRemoteLogReadResult = remoteLogReadResultArg.getValue();
        assertTrue(actualRemoteLogReadResult.error.isPresent());
        assertFalse(actualRemoteLogReadResult.fetchDataInfo.isPresent());

        // verify the record method on quota manager was called with the expected value
        ArgumentCaptor<Double> recordedArg = ArgumentCaptor.forClass(Double.class);
        verify(mockQuotaManager, times(1)).record(recordedArg.capture());
        assertEquals(0, recordedArg.getValue());

        // Verify metrics for remote reads are updated correctly
        assertEquals(1, brokerTopicStats.topicStats(TOPIC).remoteFetchRequestRate().count());
        assertEquals(0, brokerTopicStats.topicStats(TOPIC).remoteFetchBytesRate().count());
        assertEquals(1, brokerTopicStats.topicStats(TOPIC).failedRemoteFetchRequestRate().count());
        // Verify aggregate metrics
        assertEquals(1, brokerTopicStats.allTopicsStats().remoteFetchRequestRate().count());
        assertEquals(0, brokerTopicStats.allTopicsStats().remoteFetchBytesRate().count());
        assertEquals(1, brokerTopicStats.allTopicsStats().failedRemoteFetchRequestRate().count());
    }
}
